### DISCUSSION TITLE
Option to disable/increase MAX_VERSION_COUNT

### DISCUSSION THREAD

**Discussion by MDeiml** [tree-sitter/tree-sitter#1546] [Ideas]
My grammar in https://github.com/MDeiml/tree-sitter-markdown uses conflicts very heavily, as markdown is a language with a very non strict syntax. For example every `*` could be the beginning of an emphasis, but sometimes this can only be decided after parsing the whole paragraph.

This means that, to parse everything 100% correct, the `MAX_VERSION_COUNT` would need to be infinite, since emphasis can be nested. Right now it is statically set to 6 with no method to change this number.

Now it obviously makes sense to restrict the version count in most cases, even for this markdown grammar when used for highlighting for example. But it would be nice to have the option to increase or disable the maximum version count to e.g. use the grammar as a very primitive but fast markdown compiler.

How do you think about this?

**Comment by alemuller** [comment]
Can you provide a minimal example of markdown code where the limit of MAX_VERSION_COUNT is an issue?

**Reply by MDeiml** [reply]
Good suggestion. Unfortunately I could not come up with a good example. This problem only comes up in examples with very high complexity, as versions with the same parser state get combined. So an input would have to have more than 6 valid parse states at the same position. I think this happens more often when an external scanner is involved.

I've encountered some examples where this happened in the markdown parser, but I have fixed these cases by reducing some conflicts or helping the parser choose the right branches to disregard. A problem that came up for example was https://github.com/MDeiml/tree-sitter-markdown/issues/18, which was not parsed properly until including commit [ad8c32917a16dfbb387d1da567bf0c3fb6fffde2](https://github.com/MDeiml/tree-sitter-markdown/commit/ad8c32917a16dfbb387d1da567bf0c3fb6fffde2).

After playing around a bit I am no longer certain that there are actually cases where a higher MAX_VERSION_COUNT would be needed for this grammar, but even if not it might still be worth considering to provide an option.

**Reply by alemuller** [reply]
I had a look at this [markdown spec](https://github.github.com/gfm/#blocks-and-inlines). It seems it would be much easier to implement if it was broke into two. One for the blocks and another for the inline. From a quick look on your code, you almost seems to de doing this with the scanner.

Those languages with *match anything until a delimiter* are intended to be easy to parse with a hand written parser, they are note very well  suited for LR/GLR.

**Reply by MDeiml** [reply]
I thought about splitting the parser into two grammars. But I think that would be quite a bit slower as the document needs to be scanned twice.

Having just one grammar also turned out to be easier then I thought, since the problems dealing with block structure are mostly different from the problems dealing with inline structure. Conflicts for example are almost only used for inline structure.

**Reply by MDeiml** [reply]
As a side note: the gfm spec actually suggests doing two passes (which would be equivalent to two grammars).

**Reply by alemuller** [reply]
I think writing two separated parser helps writing more cohesive code, because you make stronger guarantee the rules won't be spilling over each other. Also, it is easier to maintain. And, they always be combined into a single parser.

But it also makes sense to have separated parsers  that can be shared among different flavors of markdown.

To parse, languages that lex stuff based on the context (anything until a delimiter), it is important to know *very well* how tree-sitter deals with conflicts. You'll need to write rules to create conflicts  (example bellow). , deal with white space, etc.. otherwise you will need rely on the external parser.

---

##### A (unasked) quick look at your code

First impressions.

1. The `parser.c` has more than 4Mb.
  This seems to large for a markdown parser.
2. A lot of external symbols.
  Does it is actually needed for anything other than back-ticks?
3. Lack of consistency on the style;
  It does not look good...
4. HTML;
  Can't just get the whole block and inject the html parser?
5. Validating stuff;
  Don't validate stuff on the parser.
  This can be done on the query, and it easier to maintain everything consistent.
  Things like email are pretty much impossible to validate with regex (email without "@" are legal),
  The HTML entities could be parsed by a module or by another parser, etc.

The scanner *looks* better thought. But I can't be bothered to look at it now.

Also, as pointed out, amazing test coverage. If they pass the tests and it works without hickup is good enough for me.

I hope my criticism didn't sound to harsh.

How I'd write the btick:
```js
code_span: $ => seq(
  '`',
  $._lead_btick_mark,
  // this would be another named rule, inline for clarity
  repeat(seq(
    /[^`]+/, repeat1('`') // conflict, parser calls _trail_btick_mark
  )),
  /[^`]+/,
  '`',
  $._trail_btick_mark,
),
```

This is a simple case because btick has highest precedence. But you get the idea.

And finally, "early optimization is the root of all evil". Don't worry too much about performance, tree-sitter is quite fast.

**Reply by maxbrunsfeld** [reply]
@MDeiml Awesome work on that markdown parser; it seems to work very well, judging from your test suite, and trying it out on a few of my markdown files.

I’d also be curious to see an example markdown snippet that results in many parse stack versions.

**Reply by MDeiml** [reply]
Thanks for all the interest. First of all thank you for complimenting my test coverage, but I have to give credit to the actual spec, where all test cases actually came from and also ikatyang, who implemented a previous markdown grammar from where I got my initial version of the tests.

@maxbrunsfeld getting an example that results in may parse stack versions is quite a bit more complex than I originally thought. It happens mostly with inputs where a lot of (potential) delimiter environments are open at the same time as e.g.
```
_x*x[`<a test=x_
```
But of course this is not really a good minimal example.

@alemuller 
I think it would be very interesting to split the parser into two grammars and see if the improvement in code complexity is worth it. But actually think separating my existing code more so that it is more clear what parses inline and what parses block structure would also result in a pretty well understandable code base. All in all splitting the parser and getting all test cases to work could be a lot of work for something that decreases the performance / makes actually using the grammars more complicated, so I can not guarantee that I will try this in the future. If you want to give it a try I would suggest you wait until I document / refactor my code as it's quite a mess at the moment.

Adressing your comments:
1. The `parser.c` is mostly this big because of some big regexes. For example a regex that matches all valid html entity references so just a big list of alternatives. I'm not sure if there is a better way to do this, maybe using external scanner, but I would like to avoid that if possible.
   Another big regex is the negative regex I construct for html blocks.
2. Most of these symbols come from block structure. I need to keep a stack of all open blocks to do the matching for block continuations, that is parse all the `>`s, indents etc. that belong to already open blocks. I think this is similar but more complex than parsing indentations in languages like python. I originally put all tokens that open or close blocks in the external scanner because it is hard to "communicate" between the external scanner and the rest of the grammar. I've now found a way to do so, so I'll maybe remove some external tokens in the future.
3. Jep, I'm in the process of refactoring.
4. For html blocks that is what I am doing. For inline html that is a bit more complicated, as the part relevant to decide if this is actual html can contain newlines, which are tricky.
5. Sadly markdown has to be validated by the parser. That is because every text file is valid markdown and as such it is not an option to just emit an error. I'm 100% following the spec (or at least trying to) besides one thing that is matching links. So I am actually even validating less than what the spec says. For example the spec actually says to use a specific regex for emails https://github.github.com/gfm/#email-address. This is for example important as
   ```
   *a<foo*bar>
   ```
   *a<foo*bar>
   contains emphasis whereas
   ```
   *a<foo*bar@baz.com>
   ```
   *a<foo*bar@baz.com>
   does not

As I said, I'm refactoring right now but I would love some people to review my code after that. Also should make it a lot easier to contribute once I comment everything properly

**Reply by alemuller** [reply]
I see your point, but I personally believe the Unix philosophy (*do one thing and do it well*) would be a better fit for tree-sitter for many reasons. And by doing well, in the tree-sitter parser, means doing less work. Instead of validating the HTML entities on the parser, use a injection. Same for email, just see if has an at and delegate the parsing for another parser that implements the RFC.  But, I'm not going to discuss those now.

You seems to care a lot about performance.

I avoided talking about this for a few reason:

  1. Parsers are new to me, and very outside of my comfort zone.
     My (likely wrong) uneducated guess:
     Since the worst time complexity of GLR is cubic. There may be cases where breaking the parser into two may be faster than one if you mange to reduce n.
     But this is not how asymptotic notation works. IIRC, the correct thing to say is 2*O(n^3) = O(n^3).

  2. Lack of data to support those claims;
     Haven't seem any claim that injection are slow and that it can't be made faster.

  3. There are a lot misconception of performance
     A few widely popular myths about micro arch (specifically of Intel x86):
     - Unaligned memory access are slower;
       This isn't true for a like decade.
     - Loop unrolling is good;
       Moderns CPUs have specialized hardware for small loops. Medium loops may fit in the uOP cache. Unrolling them make more likely the instructions wont fit in those. They will need to be fetch from L1 (putting pressure on the decoder).
     - Running SIMD only for a few thousands of instructions;
       The CPU takes a few *micro*seconds (~60k cycles irrc) to light up the vector engines. Meanwhile those instructions runs 1/4 of the speed.

     Of course, there are cases where those things make sense. Also, I haven't looked at the numbers from Alder Lake, but i heard it has more decoders (and no *official* avx512).

**Reply by maxbrunsfeld** [reply]
Personally, I think your current design is good. markdown is all one language; the block structure by itself isn’t really it’s own language from the user’s perspective - the two phase split is just an implementation strategy.

**Reply by alemuller** [reply]
Yeah. I agree 100%.

But. I think some things should be delegated to a specialized parsers (not only in Markdown), like:
  - HTML entity parser;
  - URI parsing;
  - Email parsing;
  - File path;
  - Escape sequence parsing (or the whole string);
  - etc

Again, not talking about markdown parser that exceptionally tries to validate them.

Having specialized parsers would lead to a more consistent highlighting in different languages, writing plugins easier, and so on.

Things like email address are not trivial to parse. Most regex doesn't conform with the RFC.

Other things like Unicode code points escape sequence are not hard, but not many people care about to implement.

Etc.

**Reply by MDeiml** [reply]
On the topic of performance I'm obviously just guessing. But at least for some use cases there is definitely some overhead for using different grammars. For example when the file is too large to fit in memory and thus must be streamed there would be two passes. One to parse the block structure and one to parse the inline structure. I am pretty sure this would be slower since often IO is a bottleneck.

I agree it would be nice to have specialized parsers for things like emails, which could also be used for markdown files using language injections. But just so we're on the same page: the markdown grammar still needs to decide whether something counts as an email or not using the regex from the spec.
